gui/seedwatch.lua

```diff
@@ -0,0 +1,310 @@
+-- config ui for seedwatch
+
+local gui = require('gui')
+local widgets = require('gui.widgets')
+local plugin = require('plugins.seedwatch')
+
+local PROPERTIES_HEADER = '   Quantity     Target   '
+local REFRESH_MS = 10000
+local MAX_TARGET = 2147483647
+--
+-- SeedSettings
+--
+SeedSettings = defclass(SeedSettings, widgets.Window)
+SeedSettings.ATTRS{
+    lockable=false,
+    frame={l=5, t=5, w=35, h=9},
+    data={id='', name='', quantity=0, target=0,},
+}
+
+function SeedSettings:init()
+    self:addviews{
+        widgets.Label{
+            frame={t=0, l=0},
+            text='Seed: ',
+        },
+        widgets.Label{
+            view_id='id',
+            frame={t=0, l=0},
+            visible=false,
+        },
+        widgets.Label{
+            view_id='name',
+            frame={t=0, l=6},
+            text_pen=COLOR_GREEN,
+        },
+        widgets.Label{
+            frame={t=1, l=0},
+            text='Quantity: ',
+        },
+        widgets.Label{
+            view_id='quantity',
+            frame={t=1, l=10},
+            text_pen=COLOR_GREEN,
+        },
+        widgets.EditField{
+            view_id='target',
+            frame={t=2, l=0},
+            label_text='Target: ',
+            key='CUSTOM_CTRL_T',
+            on_char=function(ch) return ch:match('%d') end,
+            on_submit=self:callback('commit'),
+        },
+        widgets.HotkeyLabel{
+            frame={t=4, l=0},
+            key='SELECT',
+            label='Apply',
+            on_activate=self:callback('commit'),
+        },
+    }
+end
+
+function SeedSettings:show(choice, on_commit)
+    self.data = choice.data
+    self.on_commit = on_commit
+    local data = self.data
+    self.subviews.id:setText(data.id)
+    self.subviews.name:setText(data.name)
+    self.subviews.quantity:setText(tostring(data.quantity))
+    self.subviews.target:setText(tostring(data.target))
```

Comment:
this could be simplified by working the data into function callbacks in the Labels. For example, instead of:
```lua
        widgets.Label{
            frame={t=1, l=0},
            text='Quantity: ',
        },
        widgets.Label{
            view_id='quantity',
            frame={t=1, l=10},
            text_pen=COLOR_GREEN,
        },
```
you could have
```lua
        widgets.Label{
            frame={t=1, l=0},
            text={
                'Quantity:',
                {gap=1, text=function() return self.data.quantity end, pen=COLOR_GREEN},
            },
        },
```

Reply:
done

Reply:
it can be defined at both levels, with the text (token) overriding the Label-level `text_pen`. I'm pretty sure that should work as advertised